Add nbde_client role

sergio-correia opened the pull request.

richm commented on library/nbde_client_clevis.py:
```diff
@@ -0,0 +1,1065 @@
+    cmds = []
+    cmds.append(['cryptsetup', 'luksKillSlot', '--batch-mode', device,
+                 str(slot)])
+    clevis_uuid = 'cb6e8904-81ff-40da-a84a-07ab9ab5715e'
```
This should be a global variable

sergio-correia replied:
Updated.

richm commented:
Also, please review the travis CI test failures

sergio-correia commented on library/nbde_client_clevis.py:
```diff
@@ -0,0 +1,1065 @@
+
+    if luks_type == 'luks1':
+        args = ['luksmeta', 'test', '-d', device]
+        ret, _, err = module.run_command(args)
```
Hm. It's a value we get from the playbook/variables, and it is passed to `run_command`. I am now using `shlex.quote/pipes.quote` on them during the sanity check.

sergio-correia commented on library/nbde_client_clevis.py:
```diff
@@ -0,0 +1,1065 @@
+
+    args = ['jose', 'fmt', '--json', stdout, '--object', '--get', 'jwe',
+            '--output=-']
+    ret_code, stdout, stderr = module.run_command(args)
```
Yes, it's used in the next `run_command`, a few lines down as input for another friendly `jose` command :)

sergio-correia commented:
> Also, please review the travis CI test failures

Yeah, this seems more complex to get running correctly. I would appreciate help here.

sergio-correia commented:
Thanks for the first review, @richm; I will update this shortly.

richm commented:
> > Also, please review the travis CI test failures
> 
> Yeah, this seems more complex to get running correctly. I would appreciate help here.

If you just want to skip `black` and `flake8` and `pylint` for now - see https://github.com/linux-system-roles/nbde_client/blob/master/.travis/config.sh - e.g. edit the file to set
```
export RUN_PYLINT_DISABLED=true
export RUN_BLACK_DISABLED=true
export RUN_FLAKE8_DISABLED=true
```

For black - if you want black to format your code - edit `tox.ini` like this:
```
commands =
#    bash {toxinidir}/.travis/runblack.sh --check --diff .
    bash {toxinidir}/.travis/runblack.sh .
```
then run `tox -e black` - black will reformat your python code to be conformant. Then revert `tox.ini`.

You can also review the linter issues - you can edit the various config files to completely disable that particular issue (e.g. disabling line length checking is probably the biggest one).  The config.sh variables such as `RUN_BLACK_EXTRA_ARGS` and `RUN_FLAKE8_EXTRA_ARGS` can be used to pass suppressions on the command line to those programs.

You can also use code comments to suppress specific occurrences.

sergio-correia commented:
@richm: I have removed the pytest test for now and added playbook-based tests. I will work on pytest next week. I believe this should be ready for review now.

richm commented on README.md:
```diff
@@ -1,2 +1,127 @@
+| **Name** | **Default/Choices** | **Description** |
+|----------|-------------|------|
+|`devices` | | specifies LUKS devices, with their path and possibly either a passphrase or key file.|
+|`pin` | | the settings for the binding, like the servers it should bind to. |
```
Seems like this parameter is `auth`?  It isn't really a "PIN" or even necessarily a password or pass phrase.  Is this some sort of nomenclature left over from clevis/tang?

sergio-correia replied:
Yes, it's a clevis nomenclature leftover. I think `auth` would be OK here.

richm commented on handlers/main.yml:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: MIT
+---
+# handlers file for nbde_client.
+- name: update initramfs
```
I think the handler name is in the global namespace, so it should have the role name in it e.g. `name: nbde_client update initramfs` - @pcahyna I don't think we mention this in the best practices, so perhaps I'm wrong

sergio-correia replied:
Updating this as well.

pcahyna replied:
@richm I think we have this in the best practices ...

pcahyna replied:
Sorry, we don't. At least not in the published version.

richm commented on tasks/main-clevis.yml:
```diff
@@ -0,0 +1,33 @@
+  notify: update initramfs
+
+
+- name: Ensure we dispose of any transferred keyfiles
```
Is this a task you want to perform even if `name: Perform clevis operations` fails?  If so, suggest you use a `block/rescue/always` e.g. put the cleanup in an `always` block.

sergio-correia replied:
Good point. I will do this.

richm commented on tests/cleanup_dummy_device.yml:
```diff
@@ -0,0 +1,13 @@
+---
```
Please create a `tests/tasks` sub-directory for these common task files used by the tests.
I don't think we mention this in the best practices guide but we should . . .

sergio-correia replied:
Makes sense, I will update it to follow this.

richm commented on tests/tests_bind_high_availability.yml:
```diff
@@ -0,0 +1,36 @@
+          - not nbde_unlock.failed
+          - not nbde_close.failed
+
+    - name: Clean up
```
The cleanup won't be executed if `name: Make sure the attempt to unlock succeeded` asserts

sergio-correia replied:
I will try to use the approach you suggested with the block/rescue/always. Hopefully I can get rid of the `ignore_errors: yes` as well in the verification tasks

richm replied:
> Hopefully I can get rid of the `ignore_errors: yes` as well in the verification tasks

It depends - you may actually want to run all of the verification tasks, and report all of the errors, rather than just erroring out on the first one.  It's up to you to decide what would be preferable for certain tests.  Here is an approach I took for kernel_settings - https://github.com/linux-system-roles/kernel_settings/pull/6/commits/fa71b5e9140fa2889b1336a0932b98ab105d75e6#diff-67385c2b6739d65912ec02759af26815R159

sergio-correia replied:
Okay, I am keeping with `ignore_errors` + `register` and checking the results. I am also using block/always in the tests, so that cleanup is always executed.

richm commented on README.md:
```diff
@@ -1,2 +1,127 @@
+pin:
+  slot: 3
+  servers:
+    - http://tang.server01
```
Please use RFC 5737, 7042 and 3849 addresses in examples.  https://github.com/oasis-roles/meta_standards#documentation-conventions

In tests also

richm replied:
Hmm - that doesn't seem to cover hostnames/domainnames - @pcahyna I thought we had something in the best practices about the use of hostnames/domainnames in examples and tests?

sergio-correia replied:
What about these, with `.example.com` (RFC2606)?
```
http://server1.example.com
http://server2.example.com
```

pcahyna replied:
@richm that part is not merged yet.

richm commented on tests/cleanup_dummy_device.yml:
```diff
@@ -0,0 +1,13 @@
+---
+- name: Clean up dummy testing device
+  file:
+    path: "{{ test_device }}"
```
please prefix variables used in tests with `nbde_client_`

richm replied:
One reason is that we may write integration tests with several roles in a single playbook

sergio-correia replied:
I will update this as well.

sergio-correia commented:
Thanks for the review @richm! I believe I have addressed all your points.

richm commented on tests/tasks/verify_default_keyfile.yml:
```diff
@@ -0,0 +1,9 @@
+    cryptsetup open --test-passphrase "{{ nbde_client_test_device }}"
+    --key-file "{{ nbde_client_test_keyfile}}"
+  ignore_errors: yes
+  register: nbde_keyfile
```
please use `nbde_client_keyfile`

sergio-correia replied:
Fixed this and the other occurrences. Thanks.

richm commented on tests/tasks/verify_default_keyfile.yml:
```diff
@@ -0,0 +1,9 @@
+- name: Verify the default keyfile works
+  shell: >
+    cryptsetup open --test-passphrase "{{ nbde_client_test_device }}"
+    --key-file "{{ nbde_client_test_keyfile}}"
```
please use a space before the `}}`

richm commented on tests/tasks/verify_default_passphrase.yml:
```diff
@@ -0,0 +1,9 @@
+    echo -n "{{ nbde_client_test_pass }}" |
+    cryptsetup open --test-passphrase "{{ nbde_client_test_device }}"
+  ignore_errors: yes
+  register: nbde_passphrase
```
please use `nbde_client_passphrase`

richm commented on tests/tasks/verify_unlock_device.yml:
```diff
@@ -0,0 +1,14 @@
+    clevis luks unlock -d "{{ nbde_client_test_device }}"
+    -n nbde_client_unlocked
+  ignore_errors: yes
+  register: nbde_unlock
```
please use `nbde_client_unlock`

richm commented on tests/tasks/verify_unlock_device.yml:
```diff
@@ -0,0 +1,14 @@
+- name: Close unlocked device
+  command: cryptsetup close nbde_client_unlocked
+  ignore_errors: yes
+  register: nbde_close
```
please use `nbde_client_close`

richm commented on defaults/main.yml:
```diff
@@ -0,0 +1,88 @@
+#                        than one server. When using multiple servers,
+#                        threshold indicates how many of those servers should
+#                        succeed in order to complete the decryption.
+# * overwrite          - specifies whether we should override any existing
```
Note that this is usually the default in Ansible - if you specify a value, it replaces any existing value.  Is this a case where it might be potentially very destructive to overwrite an existing binding?  e.g. there is an irreplaceable private key in a slot?

sergio-correia replied:
It should not be destructive, since we would be replacing an existing binding with another valid binding, meaning we would be able to unlock the disk with it. We can set this to `yes` as default, then.

richm replied:
I think this is the last issue - other than this, looking good

sergio-correia replied:
Thanks. This was updated, and I am now using `nbde_server` role in testing

pcahyna replied:
what happens if the value is set to `no` ?

sergio-correia replied:
It will not perform the action, if there is already a binding (output with `-v`):
```
TASK [linux-system-roles.nbde_client : Perform clevis operations] ****************************************************************************
ok: [rhel8] => {"changed": false, "msg": "/dev/rhel/root:1 is bound and no overwrite set", "original_bindings": [{"auth": {"discard_passphrase": false, "overwrite": false, "servers": ["localhost:3000"], "slot": 1, "threshold": 1}, "devices": [{"keyfile": "/root/.local/share/nbde_client/keyfile", "path": "/dev/rhel/root"}], "state": "present"}]}
```

pcahyna replied:
And it won't report an error? So the role does not do what it was told to do and pretends that everything is ok? To me this seems unusual in the context of Ansible.

sergio-correia replied:
Yeah, it seems like a bug. I will get this fixed.

pcahyna replied:
Is there an actual need for this setting? To me it seems that it could be easier to avoid aving it.

pcahyna replied:
has this been solved?

sergio-correia replied:
Yes, it's gone. Sorry, I forgot to comment here. No need to have this and it was removed.

pcahyna commented on README.md:
```diff
@@ -1,2 +1,125 @@
+|----------|-------------|------|
+|`devices` | | specifies LUKS devices, with their path and possibly either a passphrase or key file.|
+|`auth` | | the settings for the binding, like the servers it should bind to. |
+| `state` | **present** / absent | specifies whether the binding with the configuration described should be added or removed. Setting `state` to `present` (the default) means bindings will be added; setting `state` to `absent` means bindings will be removed from the devices/slots. |
```
What is the primary key to identify the binding? Is it `auth.slot`  ? it seems so, as the example below says:

#### Example 2: remove binding from slot 2
```yaml
---
- hosts: all
  vars:
    nbde_client_bindings:
      - devices:
          - path: /dev/sda1
            pass: passphrase
        auth:
          slot: 2
        state: absent
```

pcahyna replied:
Or is it the combination of (device, slot) ?

sergio-correia replied:
Yes, it's device and slot. I guess it makes sense to move slot out from auth.

pcahyna replied:
I think so, yes. Shouldn't the slot be per-device? if one binding is using two device, can the slot be different for the two? (I am really unfamiliar with nbde concepts.)

sergio-correia replied:
Each device may have bindings (more than one, even), and each binding is associated with a slot. So yeah, moving the slot to device makes sense here. We could have different devices setting the bindings in different slots.

pcahyna replied:
What happens if I run the role and add a binding with two devices and then I run the role again and I specify just one of the devices and say `state: absent`? Will the binding be split into two bindings and the one that corresponds to the device given will be removed, while the other will stay?

sergio-correia replied:
If you have two devices in there, one binding will be added in each device, in a specific slot. 
/dev/sda1 and /dev/sda2, for instance. You will have a binding in each of them, meaning that you will be able to unlock each of them (with eg. clevis), via the existing bindings.

If you afterwards specify /dev/sda2 with state absent, you will remove the binding existing in the specified device; you will not be able to unlock this one device, unless you have other bindings in other LUKS slots.

pcahyna replied:
But then it is not one binding with two devices, it is two bindings, each with one device. So the way you are representing it in the interface is inaccurate.

sergio-correia replied:
I agree. It seemed simpler to use with multiple devices, but in the end it became confusing and not as accurate.
What about something like this, instead, to describe the bindings?

```
{
	device
	pass
	keyfile
	slot
	state
	discard_passphrase
	auth {
		threshold
		servers
	}
}
```

pcahyna replied:
looks good, maybe it is not needed to group `threshold` with `servers` under `auth` and you could add those two directly in the dict?

sergio-correia replied:
Yeah, I think it would make things simpler that way. Let me update this.

sergio-correia replied:
Updated.

pcahyna commented on README.md:
```diff
@@ -1,2 +1,125 @@
+|----------|-------------|------|
+| `nbde_client_provider` | `clevis`| identifies the provider for the `nbde_client` role. We currently support `clevis`.|
+| `nbde_client_bindings` | | a list containing groups of LUKS devices and applicable settings for a binding. This group of devices contains specific binding configurations to be applied to the set of devices.
+|`nbde_client_remote_dir` | `/root/.local/share/nbde_client/` |  specifies a directory in the remote hosts that may be used for storing temporary data such as transferred keyfiles. |
```
What is a "remote host"? Is this the managed host?
If the directory is used only for temporary data, is there a need for it to be configurable?

sergio-correia replied:
Yeah, the managed host. And you are right, there is no real need for this to be configured.

pcahyna replied:
Can you please remove the variable then? We should not provide knobs that won't be useful.

sergio-correia replied:
Sure, I will fiix this.

pcahyna commented on README.md:
```diff
@@ -1,2 +1,125 @@
+| `nbde_client_provider` | `clevis`| identifies the provider for the `nbde_client` role. We currently support `clevis`.|
+| `nbde_client_bindings` | | a list containing groups of LUKS devices and applicable settings for a binding. This group of devices contains specific binding configurations to be applied to the set of devices.
+|`nbde_client_remote_dir` | `/root/.local/share/nbde_client/` |  specifies a directory in the remote hosts that may be used for storing temporary data such as transferred keyfiles. |
+|`nbde_client_update_initramfs` | `yes` | indicates whether the initramfs should be updated in case changes are made when processing the clevis operations on the devices.|
```
Is there a need for this to be configurable? Shouldn't the role figure out itself whether initramfs should be updated?

sergio-correia replied:
This is mostly useful when testing, to disable this altogether, but yeah, I guess we can remove this option.

pcahyna replied:
If you need it for development purposes, maybe you could move it from defaults to vars and prefix it with double underscores to mark it as internal.

pcahyna commented on defaults/main.yml:
```diff
@@ -0,0 +1,84 @@
+#                        binding in the specified slot.
+# * discard_passphrase - specifies whether we should discard the passphrase
+#                        provided -- via either pass or keyfile, in devices
+#                        -- after completing the binding operation.
```
what happens with the passphrase if it does not get discarded? Is it saved somewhere  ? ...

sergio-correia replied:
Nope. Actually, this is to discard the passphrase from the LUKS device, as in removing it from a keyslot in the device. In the end, you will end up with the passphrase added by the binding, and this passphrase you provided (which was valid for the device) will not be a valid passphrase anymore.

sergio-correia replied:
If it does not get discarded, we do not remove it from the LUKS keyslot

pcahyna replied:
Even with this explanation, it is hard to understand what happens, and impossible from the description given in the file or in the README. The phrase "specifies whether we should discard the passphrase provided -- via either pass or keyfile, in devices from the LUKS device, after completing the binding operation." is really hard to parse.

sergio-correia replied:
What about something like this: `discard_passphrase`: indicates whether we should discard the passphrase from the LUKS keyslot in the device. The passphrase in question is the one that was provided either via the ` pass` or `keyfile` arguments.

sergio-correia replied:
Or something with more context, like:

discard_passphrase: this option is about discarding the LUKS passphrase from the keyslot in the device, after a successful binding. Sometimes dummy passphrases are used when first creating LUKS devices, and then they are removed in post installation steps, replaced with stronger passphrases. If this option is set to yes, the passphrase that was provided via the `pass` or `keyfile` arguments will be removed from the LUKS device after the binding operation completes, i.e., they will not be valid anymore. Default is `no`.

In any case, I agree we need some rewording here.

nhosoi replied:
So, one of the typical use cases for `discard_passphrase: yes` is at the first creating LUKS device with `dummy passphrases`? Could there be any other good ones?

sergio-correia replied:
I think that may be _the_ typical case, in which we have automated (e.g kickstart) installs that set up some sort of "default" password that will be changed/removed later on. 

We suggest something like this here: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/security_hardening/configuring-automated-unlocking-of-encrypted-volumes-using-policy-based-decryption_security-hardening#configuring-automated-enrollment-of-luks-encrypted-volumes-using-kickstart_configuring-automated-unlocking-of-encrypted-volumes-using-policy-based-decryption

pcahyna replied:
> If this option is set to yes, the passphrase that was provided via the pass or keyfile arguments will be removed from the LUKS device after the binding operation completes

The README does not describe any `pass` or `keyfile` arguments!

richm commented on defaults/main.yml:
```diff
@@ -0,0 +1,48 @@
+# * servers            - specifies a list of servers to bind to. To enable
+#                        high availability, simply specify more than one server
+#                        here.
+# * threshold          - specifies the threshold for the Shamir Secret Sharing
```
Is there a default for this?  e.g. is the default value the number of servers in the `servers` list?

sergio-correia replied:
The default is 1. You can add two servers and if at least one of them is available, it will be enough to perform the decryption. If you set it to 2, both of them will be required to be available for the decryption.

richm commented on handlers/main.yml:
```diff
@@ -0,0 +1,11 @@
+  when:
+    - __nbde_client_update_initramfs | bool
+    - __nbde_client_initramfs_update_cmd is defined
+    - __nbde_client_initramfs_update_cmd != ""
```
you can combine the last two conditions as `__nbde_client_initramfs_update_cmd | d("")` - `""` should evaluate to `False` in this (boolean) context

richm commented on tasks/main-clevis.yml:
```diff
@@ -0,0 +1,30 @@
+  block:
+    - name: Ensure we transfer keyfiles
+      copy:
+        src: "{{ item.keyfile | default('/dev/null') }}"
```
What does it mean to use an empty keyfile?  Is that supported by clevis?  Should this be an error instead? e.g. if the user does not define keyfile, by default Ansible will throw an error here.

sergio-correia replied:
We need to provide a valid key for the LUKS device, via either "pass", a key file or we can even use already existing bindings and not specify either pass or keyfile. When there are keyfiles, I need to copy them to managed host, but if there are none, I can skip this (and the cleanup) tasks. I will try to use your suggestion to skip these when there is no keyfile specified.

richm commented on tasks/main-clevis.yml:
```diff
@@ -0,0 +1,30 @@
+
+    - name: Perform clevis operations
+      nbde_client_clevis:
+        bindings: "{{ nbde_client_bindings | default([]) }}"
```
Does the module actually do anything if `nbde_client_bindings` is empty?  Should we just skip the module in that case? e.g.  add a `when: nbde_client_bindings | d([])` to this task

sergio-correia replied:
Nope, it does nothing and can be skipped in that case.

sergio-correia replied:
It's being skipped now, thanks.

richm commented on tasks/main-clevis.yml:
```diff
@@ -0,0 +1,30 @@
+  always:
+    - name: Ensure we dispose of any transferred keyfiles
+      file:
+        path: "{{ __nbde_client_managed_dir }}/{{ item.keyfile | default('/dev/null') | basename }}"  # yamllint disable-line rule:line-length
```
maybe skip this task 
```yaml
  when:
    - nbde_client_bindings | d([])
    - item.keyfile | d('')
```
?

sergio-correia replied:
Yeah, this should be perfect, to skip the task if there are no keyfiles to copy. Let me try this.

sergio-correia replied:
Skipping here and in the copy task also, thanks.

richm commented on tests/tests_discard_passphrase.yml:
```diff
@@ -0,0 +1,45 @@
+        - name: Make sure the default passphrase did not work
+          assert:
+            that:
+              - nbde_client_passphrase.rc != 0
```
since `nbde_client_passphrase` is a register variable, use `- nbde_client_passphrase is success` rather than checking `.rc` directly - see https://docs.ansible.com/ansible/latest/user_guide/playbooks_tests.html#task-results

richm commented on tests/tests_discard_passphrase_keyfile.yml:
```diff
@@ -0,0 +1,45 @@
+        - name: Make sure the default keyfile did not work
+          assert:
+            that:
+              - nbde_client_keyfile.rc != 0
```
use `is success` similar to https://github.com/linux-system-roles/nbde_client/pull/3/files#r437623134

richm commented:
ok - lgtm - @pcahyna have your concerns been addressed?
looks like there is a yamllint error now

richm commented:
got staging CI working - any idea why the Fedora tests are failing?

sergio-correia commented:
> got staging CI working - any idea why the Fedora tests are failing?

It's using the master version of the nbde_server role. The issue is probably due to the (lack of) cache update.

nhosoi commented on README.md:
```diff
@@ -1,2 +1,98 @@
+| `device` | | specifies the path of a LUKS device (**REQUIRED**). |
+| `state` | **present** / absent | specifies whether a binding with the configuration described should be added or removed. Setting state to present (the default) means a binding will be added; setting state to absent means a binding will be removed from the device/slot. |
+| `slot` | `1` | specifies the slot to use for the binding. |
+| `servers` | |  specifies a list of servers to bind to. To enable high availability, simply specify more than one server here. |
```
(**REQUIRED**)?

sergio-correia replied:
Right, it's not needed for `absent`

pcahyna replied:
I would delete "simply".

nhosoi commented on library/nbde_client_clevis.py:
```diff
@@ -0,0 +1,1246 @@
+def unbind_slot_luks2(module, device, slot):
+    """ Unbind slot in a LUKS2 device. This involves removing both the clevis
+    metadata as well as its associated keyslot.
+    Retrurn <result> <error> """
```
s/Retrurn/Return/

nhosoi commented on library/nbde_client_clevis.py:
```diff
@@ -0,0 +1,1246 @@
+    Return <result> <error> """
+    _, err = get_jwe_luks1(module, device, slot)
+    if err:
+        errmsg = "{}:{} is not bound to clevs".format(device, slot)
```
clevs? clevis?

nhosoi commented on library/nbde_client_clevis.py:
```diff
@@ -0,0 +1,1246 @@
+    Retrurn <result> <error> """
+    _, token_id, err = get_jwe_luks2(module, device, slot)
+    if err:
+        errmsg = "{}:{} is not bound to clevs".format(device, slot)
```
clevs? clevis?

richm commented:
Does this look good to everyone?  All we need is better text for the `passphrase_temporary`?

pcahyna commented on library/nbde_client_clevis.py:
```diff
@@ -0,0 +1,1246 @@
+          pass: password
+      state: absent
+      auth:
+        slot: 2
```
Both examples are obsolete.

pcahyna commented:
> Does this look good to everyone? All we need is better text for the `passphrase_temporary`?

First of all, the code and examples need changing, they are still using `discard_passphrase`

pcahyna commented on README.md:
```diff
@@ -1,2 +1,98 @@
+
+
+#### nbde_client_bindings
+`nbde_client_bindings` supports the following keys:
```
I would say "is a list of dictionaries that support the following keys"

pcahyna commented on defaults/main.yml:
```diff
@@ -0,0 +1,47 @@
+# * device             - specifies the path of a LUKS device. (REQUIRED)
+# * pass               - specifies a valid passphrase for this device.
+# * keyfile            - specifies either the absolute or relative path of a
+#                        valid keyfile for the device pointed out by device.
```
On the managed host or on the control node?

sergio-correia replied:
> > Does this look good to everyone? All we need is better text for the `passphrase_temporary`?
> 
> First of all, the code and examples need changing, they are still using `discard_passphrase`

discard_passphrase was not changed yet, so this is expected, but the examples in the library obsolete, as you pointed out. Let me update them.

sergio-correia replied:
> On the managed host or on the control node?

on the control node. Updating this as well.

pcahyna replied:
> > On the managed host or on the control node?
> 
> on the control node. Updating this as well.

That's too bad, because the same parameter in the storage role specifies a path on the managed node... https://github.com/linux-system-roles/storage/pull/102#discussion_r439003264 is there a good reason for this difference?

sergio-correia replied:
> > > On the managed host or on the control node?
> > 
> > 
> > on the control node. Updating this as well.
> 
> That's too bad, because the same parameter in the storage role specifies a path on the managed node... [linux-system-roles/storage#102 (comment)](https://github.com/linux-system-roles/storage/pull/102#discussion_r439003264) is there a good reason for this difference?

Actually, for us the `device` is on the managed host and the `keyfile` is on the control node, and we would copy it ourselves. From your comment, I thought you were referring to the `keyfile`.

I don't feel strongly about requiring the keyfile already to be on the managed host. In fact, it would make things simpler for us. @richm, what do you think here?

pcahyna replied:
(disclaimer: quite possibly on the control node makes more sense, I just dislike the inconsistency in semantics.)

richm replied:
> I don't feel strongly about requiring the keyfile already to be on the managed host. In fact, it would make things simpler for us. @richm, what do you think here?

How will the file get to the `keyfile` location on the managed host?

sergio-correia replied:
If we require it to be on the managed host, it would be the admin's responsibility to get it there. I think what we are doing now (copying it ourselves) makes sense.

pcahyna replied:
@sergio-correia 
> From your comment, I thought you were referring to the keyfile

Yes, I was referring to the keyfile.

> I don't feel strongly about requiring the keyfile already to be on the managed host. In fact, it would make things simpler for us. 

Important is what is simpler for the user, and I have no idea how keyfiles are used in practice. Are they usually lying on the managed host? By the way, in storage we seem to be leaning towards having keyfile on the control node as well. Can you please help them with the implementation?

pcahyna commented on README.md:
```diff
@@ -1,2 +1,98 @@
+| `slot` | `1` | specifies the slot to use for the binding. |
+| `servers` | |  specifies a list of servers to bind to. To enable high availability, specify more than one server here. |
+| `threshold` | `1` | specifies the threshold for the Shamir Secret Sharing (SSS) scheme that is put in place when using more than one server. When using multiple servers, threshold indicates how many of those servers should succeed, in terms of decryption, in order to complete the process of recovering the LUKS passphrase to open the device. |
+| `discard_passphrase` | `no` | specifies whether we should discard the passphrase provided -- via either pass or keyfile, in devices from the LUKS device, after completing the binding operation. |
```
If `yes`, the passphrase that was provided via the `pass` or `keyfile` arguments will be used to unlock the LUKS device and then it will be removed from the LUKS device after the binding operation completes, i.e., it will not be valid anymore. To be used if `device` has been previously created with a dummy passphrase (for example by an automated  install like kickstart that set up some sort of "default" password), which the role should replace by a stronger one.

pcahyna commented on README.md:
```diff
@@ -1,2 +1,98 @@
+
+  vars:
+    nbde_client_bindings:
+      - device: /tmp/sda1
```
um, `/tmp/sda` ?

pcahyna commented on README.md:
```diff
@@ -1,2 +1,98 @@
+`nbde_client_bindings` supports the following keys:
+| **Name** | **Default/Choices** | **Description** |
+|----------|-------------|------|
+| `device` | | specifies the path of a LUKS device (**REQUIRED**). |
```
"LUKS must be already set up on this device before using the role" or "this device must be already configured as a LUKS device before using the role" or something similar.

pcahyna commented on README.md:
```diff
@@ -1,2 +1,100 @@
+`nbde_client_bindings` is a list of dictionaries that support the following keys:
+| **Name** | **Default/Choices** | **Description** |
+|----------|-------------|------|
+| `device` | | specifies the path of the backing of an encrypted device in the managed host. This device must be already configured as a LUKS device before using the role (**REQUIRED**). |
```
"of the backing" - backing what? shouldn't it read "of the backing device"?

pcahyna replied:
@dwlehman what do you think?

sergio-correia replied:
I updated this to use "of the backing device"

pcahyna commented on README.md:
```diff
@@ -1,2 +1,100 @@
+|----------|-------------|------|
+| `device` | | specifies the path of the backing of an encrypted device in the managed host. This device must be already configured as a LUKS device before using the role (**REQUIRED**). |
+| `passphrase` | | a valid passphrase for opening/unlocking the specified device. |
+| `keyfile` | | either the absolute or relative path, in the control node, of a keyfile valid for opening/unlocking the specified device. |
```
`key_file`, for consistency with storage. Also, I believe "on the control node" looks better.

pcahyna commented on README.md:
```diff
@@ -1,2 +1,100 @@
+| `slot` | `1` | specifies the slot to use for the binding. |
+| `servers` | |  specifies a list of servers to bind to. To enable high availability, specify more than one server here. |
+| `threshold` | `1` | specifies the threshold for the Shamir Secret Sharing (SSS) scheme that is put in place when using more than one server. When using multiple servers, threshold indicates how many of those servers should succeed, in terms of decryption, in order to complete the process of recovering the LUKS passphrase to open the device. |
+| `passphrase_temporary` | `no` | If yes, the passphrase that was provided via the pass or keyfile arguments will be used to unlock the device and then it will be removed from the LUKS device after the binding operation completes, i.e. it will not be valid anymore. To be used if device has been previously created with a dummy passphrase (for example by an automated install like kickstart that set ...
```
s/pass/passphrase/
s/keyfile/key_file/

pcahyna commented on README.md:
````diff
@@ -1,2 +1,100 @@
+```yaml
+---
+- hosts: all
+
````
I believe you can remove the empty line, or is that the usual style of Ansible docs?

sergio-correia replied:
Looking at https://docs.ansible.com/ansible/latest/user_guide/playbooks_intro.html I see both with and without this line space in the examples -- not sure there is a convention.

I checked now and saw that we don't have this empty line in the example for the kernel_settings role, for instance, so I removed it for consistency, in any case.

pcahyna commented on README.md:
````diff
@@ -1,2 +1,100 @@
+```yaml
+---
+- hosts: all
+
````
ditto

pcahyna commented on defaults/main.yml:
```diff
@@ -0,0 +1,57 @@
+#                        as a LUKS device before using the role (REQUIRED).
+# * pass               - specifies a valid passphrase for opening/unlocking
+#                        the specified device.
+# * keyfile            - specifies either the absolute or relative path, in
```
key_file

pcahyna commented on defaults/main.yml:
```diff
@@ -0,0 +1,57 @@
+# * device             - specifies the path of an encrypted device in the
+#                        managed host. This device must be already configured
+#                        as a LUKS device before using the role (REQUIRED).
+# * pass               - specifies a valid passphrase for opening/unlocking
```
passphrase

pcahyna commented on defaults/main.yml:
```diff
@@ -0,0 +1,57 @@
+#      passphrase_temporary: no
+#      servers:
+#        - http://server1.example.com
+#        - http://server2.example.com
```
Is it necessary to reproduce the whole reference documentation here? I think it would be enough to describe the variable in a single sentence. @richm what do you think?

sergio-correia replied:
That's what I was going to ask. I think a single sentence would suffice here, and the README is the actual documentation.

richm replied:
Almost none of the other system roles have extensive examples in `defaults/main.yml` (I'm guilty of this in kernel_settings to some extent :-| )

pcahyna commented on tasks/main-clevis.yml:
```diff
@@ -0,0 +1,39 @@
+  block:
+    - name: Ensure we transfer keyfiles
+      copy:
+        src: "{{ item.keyfile }}"
```
`key_file`

pcahyna commented on tasks/main-clevis.yml:
```diff
@@ -0,0 +1,39 @@
+        dest: "{{ __nbde_client_managed_dir }}/"
+        mode: '0400'
+      when:
+        - item.keyfile | default("")
```
`key_file`

pcahyna commented on tasks/main-clevis.yml:
```diff
@@ -0,0 +1,39 @@
+        - item.keyfile | default("")
+      loop: "{{ nbde_client_bindings }}"
+      loop_control:
+        label: "{{ item.keyfile | default('') }}"
```
key_file

pcahyna commented on tasks/main-clevis.yml:
```diff
@@ -0,0 +1,39 @@
+- name: Prepare keyfiles, perform clevis operations and dispose of keyfiles
+  block:
+    - name: Ensure we transfer keyfiles
+      copy:
```
the `tempfile` module could be useful. This way you would not have to define a __nbde_client_managed_dir . https://docs.ansible.com/ansible/latest/modules/tempfile_module.html#tempfile-module

sergio-correia replied:
Thanks, I will look into using it instead.

richm replied:
Here is an example from tlog - https://github.com/linux-system-roles/tlog/blob/master/tests/tests_default_wrapper.yml - you create the `tempfile` in a `block`, then use the `always` clause to remove it.

sergio-correia replied:
Thanks, I am using `tempfile` now.

richm commented on tasks/main-clevis.yml:
```diff
@@ -0,0 +1,45 @@
+  package:
+    name: "{{ __nbde_client_packages }}"
+    state: present
+
```
extra newline here

richm commented on tests/tasks/setup_test.yml:
```diff
@@ -0,0 +1,50 @@
+---
+- name: Include general tests variables
+  include_vars: main.yml
+
```
prefer only 1 empty line between tasks instead of 2

sergio-correia replied:
is there a way to enforce this in yamllint?

richm replied:
> is there a way to enforce this in yamllint?

not sure - will investigate

richm replied:
hmm - I don't think this is a yaml thing, I think it is an ansible thing - https://docs.ansible.com/ansible-lint/ doesn't seem to have a check for that @pcahyna @tyll do you know if there is a linter for this?  OTOH, I don't recall if 'one blank line between tasks' is in any of the best practices, but it is a common convention.

tyll replied:
I hear about this for the first time. Would be great to have a linter or actually formatter for this.

richm commented on tasks/main-clevis.yml:
```diff
@@ -0,0 +1,45 @@
+    state: present
+
+
+- name: Prepare key files, perform clevis operations and dispose of key files
```
Is this role idempotent?  If you run the role again with the same `nbde_client_bindings`, will it do all of these steps again, and report that something changed?  If so, I'm not sure how to prevent that from happening - how would you know ahead of time that you have already brought the managed hosts to the desired state?  I realize it may be a little late for this particular PR, and I apologize for not noticing this sooner, so perhaps we can address this in a subsequent PR if necessary.

sergio-correia replied:
Not fully. We discovered an issue with at least `passphrase_temporary`, which will be addressed after the initial merge.

sergio-correia replied:
@pcahyna opened #4 to track this issue.

richm commented on README.md:
```diff
@@ -1,2 +1,98 @@
+    state: present
+    slot: 2
+    threshold: 1
+    passphrase_etmporary: no
```
passphrase_temporary

richm commented on README.md:
```diff
@@ -1,2 +1,98 @@
+| **Name** | **Default/Choices** | **Description** |
+|----------|-------------|------|
+| `device` | | specifies the path of the backing device of an encrypted device on the managed host. This device must be already configured as a LUKS device before using the role (**REQUIRED**). |
+| `passphrase` | | a valid passphrase for opening/unlocking the specified device. |
```
If this needs to be protected in the Ansible inventory, we should include some text like this:
```
recommend vault encrypting the private key password
see https://docs.ansible.com/ansible/latest/user_guide/vault.html
```
As in the network docs - https://github.com/linux-system-roles/network/#examples-of-options

richm replied:
> If this needs to be protected in the Ansible inventory, we should include some text like this:
> 
> ```
> recommend vault encrypting the passphrase
> see https://docs.ansible.com/ansible/latest/user_guide/vault.html
> ```
> 
> As in the network docs - https://github.com/linux-system-roles/network/#examples-of-options

sergio-correia replied:
This is a good idea.
